src/skagent/bellman.py

```diff
@@ -12,88 +12,61 @@
 import torch
 from torch.autograd import grad
 
-
-def create_transition_function(block, state_syms, decision_rules=None):
+class BellmanPeriod():
     """
-    block
-    state_syms : list of string
-        A list of symbols for 'state variables at time t', aka arrival states.
-        # TODO: state variables should be derived from the block analysis.
+    A class representing a period of a Bellman or Dynamic Stochastic Optimization Problem.
+
+    TODO: Currently this is based on a Block, but I think BlockBellmanPeriod should be
+    a subclass of an abstract class.
     """
-    decision_rules = {} if decision_rules is None else decision_rules
 
-    def transition_function(states_t, shocks_t, controls_t, parameters):
-        vals = parameters | states_t | shocks_t | controls_t
-        post = block.transition(vals, decision_rules, fix=list(controls_t.keys()))
+    def __init__(self, block, calibration, decision_rules = None):
+        self.block = block
+        self.calibration = calibration
+        self.decision_rules = decision_rules
+        self.arrival_states = self.block.get_arrival_states(calibration)
 
-        return {sym: post[sym] for sym in state_syms}
+    def get_arrival_states(self, calibration):
+        return self.block.get_arrival_states(calibration if calibration else self.calibration)
 
-    return transition_function
+    def get_controls(self):
+        return self.block.get_controls()
 
+    def get_shocks(self):
+        return self.block.get_shocks()
 
-def create_decision_function(block, decision_rules):
-    """
-    block
-    decision_rules
-    """
+    def transition_function(self, states_t, shocks_t, controls_t, parameters = None, decision_rules = None):
+        decision_rules = decision_rules if decision_rules else (self.decision_rules if self.decision_rules else {})
+        parameters = parameters if parameters else (self.calibration if self.calibration else {})
+
+        vals = parameters | states_t | shocks_t | controls_t
+        post = self.block.transition(vals, decision_rules, fix=list(controls_t.keys()))
```

Comment:
I like this, a functional transition that takes well-separated inputs $f(x, \epsilon, a; \cdot)$